dot/state/grandpa.go

```diff
@@ -24,44 +26,570 @@ var (
+		}
+	}
+
+	pendingChangeNode := &pendingChangeNode{
+		change: pendingChange,
+		nodes:  []*pendingChangeNode{},
+	}
+
+	s.scheduledChangeRoots = append(s.scheduledChangeRoots, pendingChangeNode)
+	logger.Debugf("there are now %d possible scheduled changes (roots)", len(s.scheduledChangeRoots))
+
+	return nil
+}
+
+// getApplicableScheduledChange iterates through the scheduled change tree roots looking
+// for the change node which contains a lower or equal effective number, to apply.
+// When we found that node we update the scheduled change tree roots with its children
+// that belongs to the same finalized node branch. If we don't find such node we update the scheduled
+// change tree roots with the change nodes that belongs to the same finalized node branch
+func (s *GrandpaState) getApplicableScheduledChange(finalizedHash common.Hash, finalizedNumber uint) (
+	change *pendingChange, err error) {
+
+	var changeNode *pendingChangeNode
+	for _, root := range s.scheduledChangeRoots {
+		if root.change.effectiveNumber() > finalizedNumber {
+			continue
+		}
+
+		changeNodeHash := root.change.announcingHeader.Hash()
+
+		// if the change doesn't have the same hash
+		// neither the finalized header is descendant of the change then skip to the next root
+		if !finalizedHash.Equal(changeNodeHash) {
+			isDescendant, err := s.blockState.IsDescendantOf(changeNodeHash, finalizedHash)
+			if err != nil {
+				return nil, fmt.Errorf("cannot verify ancestry: %w", err)
+			}
+
+			if !isDescendant {
+				continue
+			}
+		}
+
+		// the changes must be applied in order, so we need to check if our finalized header
+		// is ahead of any children, if it is that means some previous change was not applied
+		for _, child := range root.nodes {
+			isDescendant, err := s.blockState.IsDescendantOf(child.change.announcingHeader.Hash(), finalizedHash)
+			if err != nil {
+				return nil, fmt.Errorf("cannot verify ancestry: %w", err)
+			}
+
+			if child.change.announcingHeader.Number <= finalizedNumber && isDescendant {
+				return nil, errUnfinalizedAncestor
+			}
+		}
+
+		changeNode = root
+		break
+	}
+
+	// if there is no change to be applied then we should keep only
+	// the scheduled changes which belongs to the finalized header
+	// otherwise we should update the scheduled roots to be the child
+	// nodes of the applied scheduled change
+	if changeNode == nil {
+		err := s.keepDescendantScheduledChanges(finalizedHash)
+		if err != nil {
+			return nil, fmt.Errorf("cannot keep descendant scheduled nodes: %w", err)
+		}
+	} else {
+		change = changeNode.change
+		s.scheduledChangeRoots = make([]*pendingChangeNode, len(changeNode.nodes))
+		copy(s.scheduledChangeRoots, changeNode.nodes)
+	}
+
+	return change, nil
+}
+
+// keepDescendantForcedChanges should keep the forced changes for later blocks that
+// are descendant of the finalized block
+func (s *GrandpaState) keepDescendantForcedChanges(finalizedHash common.Hash, finalizedNumber uint) error {
```

Comment:
you should rename to `pruneForcedChanges`, since it doesn't really `keep`, it `prunes` I would say :thinking: